### ISSUE TITLE
Generate fewer specialized states

### ISSUE THREAD

**Issue by apt1002** [rrthomas/smite#257]
We observe that when the specializer training data includes many repeats of a long instruction sequence, the number of states required grows as the square of the length of the sequence. Intuitively a linear growth is sufficient.

The pForth function prologue and epilogue are such sequences. They are only a few dozen instructions long, yet there are thousands of states dedicated to them, and no states dedicated to anything else.

The design bug is that we create a state for every frequent instruction sequence, including those that only occur as part of a longer frequent sequence. In practice, the only states that are used are those corresponding to prefixes of the long sequence. The rest would only be used if the interpreter were started in the middle of the long sequence. We would prefer these other states not to exist.

One idea is to generate a state for a sequence `S` only if `xS` is frequent for at least two distinct instructions `x`. There might be a better idea.

**Comment by rrthomas** [comment]
Dumb question: it seems to me that cake-and-eat-it we want to generate states for frequent sequences and not their subsequences. At the moment I guess we can't directly represent this, since we represent sequences without noting where they come from. So if the prolog sequence is ABCDEF and occurs 10,000 times, we don't know that of the 10,004 times that ABCD appears, 10,000 are as part of ABCDEF.

So I'm wondering whether there's a way to collect this information from the trace so that one can extract the "real" sequences: LZ78-compress the trace, counting how often each dictionary entry is used, and then use the most-frequently-used entries as the states.

**Comment by apt1002** [comment]
We do know (approximately) how often each sequence occurs. So if ABCDEF occurs 10000 times and ABCD occurs 10004 times, we've explained all but 4. This is definitely a viable way of computing the frequencies of prefixes and suffixes.

An important requirement is that all the states are reachable from the initial state, which corresponds to the empty sequence. "Reachable" means by adding suffixes and removing prefixes, one instruction at a time. That's why I said we need states corresponding to prefixes of the frequent long sequences.

A case to think about is a tight loop. We want states corresponding (roughly) to all rotations of the loop body, but we want all prefixes of at most one of those states. (The prefixes are only needed if the interpreter starts inside the loop).

**Comment by apt1002** [comment]
## Sub-problem 1

I think we should generate a state for sequence `S` only if `freq(S) >= T` and there does not exist a non-empty prefix `P` such that `freq(PS) >= T` and `freq(PS) >= freq(S) - T`. In this definition, `T` is a tuning parameter indicating how frequent a sequence has to be to consider it at all. I have reused `T` to test whether there is a significant different between `S` and `PS`; alternatively an additional tuning parameter could be used for that purpose. Note that short (frequent) prefixes `P` tend to be the ones most likely to disqualify `S`.

## Sub-problem 2

In practice, we don't want to have to specify `T`. We would prefer to specify the number of states. Thus there is a separable problem: how to find `T` and compute the set efficiently. I think one way is as follows:

 - Start `T` at infinity and gradually reduce it during the algorithm until the output set is big enough.
 - Maintain a priority queue of events, ordered by the `T` value at which the event occurs. The queue contains two kinds of event. Initially, the queue contains only a type-1 event for the empty sequence, where the two types are:
    1. `T` reduces below `freq(S)` for some `S`.
    2. `T` reduces below `freq(S) - freq(PS)` for some `S` and its most common prefix `P`.
 - When a type-1 event occurs for some string `S`, add it to the output set, and for each opcode `x` insert a type-1 event for `Sx`.
 - When a type-1 event occurs for some string `PS`, remove `S` from the output set, and insert a type-2 event for `S`, if one does not already exist.
 - When a type-2 event occurs for `S`, add it back to the output set.

## Strategy

I suggest we focus on sub-problem 1 at first, using a manually configured `T`, and see how the specializer behaves. Only if we're happy with the solution is it worth investing in sub-problem 2.

There is an interaction with `State.is_worthwhile()`; sometimes, we encounter strings `P` and `S` such that `PS` is worthwhile and `S` is not. There is also an interaction with `Path.remove_repeating_part()`. It would be nice to separate these concerns out somehow.

**Comment by apt1002** [comment]
On branch `fewer-states-257` we tried "sub-problem 1" above. It builds much faster, and the number of states has reduced. After resolving some confusion, we broke our record for the time needed to build pForth.

We didn't do sub-problem 2. Instead, we generated a fixed number of paths and then deleted most of them. This results in a reasonable (and controlled) if slightly unpredictable number of states.
